From c2d7c3f5bf38481df25e30ba8457a3e8b46a228e Mon Sep 17 00:00:00 2001
From: Wanwei Jiang <wanwei.jiang@amlogic.com>
Date: Thu, 2 Feb 2023 15:49:10 +0800
Subject: [PATCH] init: initialize the building framework of android-mainline
 for amlogic [1/1]

PD#SWPL-109858

Problem:
Initialize the building framework of android-mainline for amlogic

Solution:
Initialize the building framework of android-mainline for amlogic

Verify:
sc2

Change-Id: Icfbc29709af1545b802ed7794808a2ab560ed08a
Signed-off-by: Wanwei Jiang <wanwei.jiang@amlogic.com>
Signed-off-by: Dezhen Wang <dezhen.wang@amlogic.com>
---
 .gitignore               |   3 +
 Kbuild                   |   9 +++
 Makefile                 |  44 ++++++++++-
 scripts/checkpatch.pl    | 158 +++++++++++++++++++--------------------
 scripts/kconfig/Makefile |   9 ++-
 usr/include/Makefile     |   4 +
 6 files changed, 146 insertions(+), 81 deletions(-)

diff --git a/.gitignore b/.gitignore
index a61e4778d011..b0f871c07ad4 100644
--- a/.gitignore
+++ b/.gitignore
@@ -178,3 +178,6 @@ sphinx_*/
 
 # Rust analyzer configuration
 /rust-project.json
+
+# amlogic drivers
+/common_drivers/
diff --git a/Kbuild b/Kbuild
index 464b34a08f51..88e1b7633843 100644
--- a/Kbuild
+++ b/Kbuild
@@ -97,3 +97,12 @@ obj-$(CONFIG_SAMPLES)	+= samples/
 obj-$(CONFIG_NET)	+= net/
 obj-y			+= virt/
 obj-y			+= $(ARCH_DRIVERS)
+ifeq ($(CONFIG_AMLOGIC_IN_KERNEL_MODULES),y)
+ifeq ($(CONFIG_AMLOGIC_DDK_BUILD),y)
+	obj-y		+= $(COMMON_DRIVERS_DIR)/drivers/tty/serial/
+else
+	obj-y		+= $(COMMON_DRIVERS_DIR)/drivers/
+	obj-y		+= $(COMMON_DRIVERS_DIR)/sound/
+	obj-y		+= $(COMMON_DRIVERS_DIR)/samples/
+endif
+endif
diff --git a/Makefile b/Makefile
index dac24d2162c2..650a2730c500 100644
--- a/Makefile
+++ b/Makefile
@@ -428,6 +428,20 @@ ifneq ($(SRCARCH),$(SUBARCH))
 cross_compiling := 1
 endif
 
+# ifdef CONFIG_AMLOGIC_DRIVER
+# config cannot be used here to mark AMLOGIC modifications
+# If these three variables are not set externally, set their default values
+ifeq ($(shell test -d $(srctree)/common_drivers; echo $$?),0)
+       export COMMON_DRIVERS_DIR := common_drivers
+else
+       export COMMON_DRIVERS_DIR := ../common_drivers
+endif
+export KCONFIG_EXT_PREFIX ?= ${COMMON_DRIVERS_DIR}/
+export KCONFIG_PROJECT_PREFIX ?= ${COMMON_DRIVERS_DIR}/
+export dtstree ?= $(COMMON_DRIVERS_DIR)/arch/$(SRCARCH)/boot/dts/
+export DTC_INCLUDE ?= $(srctree)/$(COMMON_DRIVERS_DIR)/include
+# endif
+
 KCONFIG_CONFIG	?= .config
 export KCONFIG_CONFIG
 
@@ -833,6 +847,19 @@ endif # need-config
 
 KBUILD_CFLAGS	+= -fno-delete-null-pointer-checks
 
+ifdef CONFIG_AMLOGIC_DRIVER
+USERINCLUDE    += \
+		-I$(srctree)/$(COMMON_DRIVERS_DIR)/include \
+		-I$(srctree)/$(COMMON_DRIVERS_DIR)/include/uapi
+
+LINUXINCLUDE   += \
+		-I$(srctree)/$(COMMON_DRIVERS_DIR)/include \
+		-I$(srctree)/$(COMMON_DRIVERS_DIR)/include/uapi
+KBUILD_CFLAGS += -Werror
+KBUILD_CFLAGS += -Wno-override-module
+KBUILD_AFLAGS += -Wno-unused-command-line-argument
+endif
+
 ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE
 KBUILD_CFLAGS += -O2
 KBUILD_RUSTFLAGS += -Copt-level=2
@@ -1184,6 +1211,9 @@ ifeq ($(KBUILD_EXTMOD),)
 endif
 	$(Q)$(MAKE) $(hdr-inst)=$(hdr-prefix)include/uapi
 	$(Q)$(MAKE) $(hdr-inst)=$(hdr-prefix)arch/$(SRCARCH)/include/uapi
+ifdef CONFIG_AMLOGIC_DRIVER
+	$(Q)$(MAKE) $(hdr-inst)=$(hdr-prefix)$(COMMON_DRIVERS_DIR)/include/uapi
+endif
 
 ifeq ($(KBUILD_EXTMOD),)
 
@@ -1453,10 +1483,22 @@ endif
 ifneq ($(dtstree),)
 
 %.dtb: dtbs_prepare
+ifdef CONFIG_AMLOGIC_DRIVER
+	$(if $(wildcard $(srctree)/$(dtstree)/amlogic/$(addsuffix .dts,$(basename $@))),\
+		$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/amlogic/$@,\
+		$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/$@)
+else
 	$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/$@
+endif
 
 %.dtbo: dtbs_prepare
+ifdef CONFIG_AMLOGIC_DRIVER
+	$(if $(wildcard $(srctree)/$(dtstree)/amlogic/$(addsuffix .dts,$(basename $@))),\
+		$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/amlogic/$@,\
+		$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/$@
+else
 	$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/$@
+endif
 
 PHONY += dtbs dtbs_prepare dtbs_install dtbs_check
 dtbs: dtbs_prepare
diff --git a/scripts/checkpatch.pl b/scripts/checkpatch.pl
index b03d526e4c45..8c7451e4a444 100755
--- a/scripts/checkpatch.pl
+++ b/scripts/checkpatch.pl
@@ -60,7 +60,7 @@ my $configuration_file = ".checkpatch.conf";
 my $max_line_length = 100;
 my $ignore_perl_version = 0;
 my $minimum_perl_version = 5.10.0;
-my $min_conf_desc_length = 4;
+my $min_conf_desc_length = 2;
 my $spelling_file = "$D/spelling.txt";
 my $codespell = 0;
 my $codespellfile = "/usr/share/codespell/dictionary.txt";
@@ -460,8 +460,8 @@ if ($tree) {
 	}
 
 	if (!defined $root) {
-		print "Must be run from the top-level dir. of a kernel tree\n";
-		exit(2);
+		# print "Must be run from the top-level dir. of a kernel tree\n";
+		# exit(2);
 	}
 }
 
@@ -1151,7 +1151,7 @@ sub is_maintained_obsolete {
 sub is_SPDX_License_valid {
 	my ($license) = @_;
 
-	return 1 if (!$tree || which("python3") eq "" || !(-x "$root/scripts/spdxcheck.py") || !(-e "$gitroot"));
+	return 1 if (!$tree || which("python3") eq "" || !(-x "scripts/spdxcheck.py") || !(-e "$gitroot"));
 
 	my $root_path = abs_path($root);
 	my $status = `cd "$root_path"; echo "$license" | scripts/spdxcheck.py -`;
@@ -2868,7 +2868,7 @@ sub process {
 
 			$p1_prefix = $1;
 			if (!$file && $tree && $p1_prefix ne '' &&
-			    -e "$root/$p1_prefix") {
+			    -e "$p1_prefix") {
 				WARN("PATCH_PREFIX",
 				     "patch prefix '$p1_prefix' exists, appears to be a -p0 patch\n");
 			}
@@ -2892,10 +2892,10 @@ sub process {
 		}
 
 		if ($found_file) {
-			if (is_maintained_obsolete($realfile)) {
-				WARN("OBSOLETE",
-				     "$realfile is marked as 'obsolete' in the MAINTAINERS hierarchy.  No unnecessary modifications please.\n");
-			}
+			# if (is_maintained_obsolete($realfile)) {
+			# 	WARN("OBSOLETE",
+			# 	     "$realfile is marked as 'obsolete' in the MAINTAINERS hierarchy.  No unnecessary modifications please.\n");
+			# }
 			if ($realfile =~ m@^(?:drivers/net/|net/|drivers/staging/)@) {
 				$check = 1;
 			} else {
@@ -2910,8 +2910,8 @@ sub process {
 
 				if (($last_binding_patch != -1) &&
 				    ($last_binding_patch ^ $is_binding_patch)) {
-					WARN("DT_SPLIT_BINDING_PATCH",
-					     "DT binding docs and includes should be a separate patch. See: Documentation/devicetree/bindings/submitting-patches.rst\n");
+					# WARN("DT_SPLIT_BINDING_PATCH",
+					#      "DT binding docs and includes should be a separate patch. See: Documentation/devicetree/bindings/submitting-patches.rst\n");
 				}
 			}
 
@@ -2933,9 +2933,9 @@ sub process {
 				$commit_log_lines++;	#could be a $signature
 			}
 		} elsif ($has_commit_log && $commit_log_lines < 2) {
-			WARN("COMMIT_MESSAGE",
-			     "Missing commit description - Add an appropriate one\n");
-			$commit_log_lines = 2;	#warn only once
+			# WARN("COMMIT_MESSAGE",
+			#      "Missing commit description - Add an appropriate one\n");
+			# $commit_log_lines = 2;	#warn only once
 		}
 
 # Check if the commit log has what seems like a diff which can confuse patch
@@ -3073,8 +3073,8 @@ sub process {
 			my ($email_name, $name_comment, $email_address, $comment) = parse_email($email);
 			my $suggested_email = format_email(($email_name, $name_comment, $email_address, $comment));
 			if ($suggested_email eq "") {
-				ERROR("BAD_SIGN_OFF",
-				      "Unrecognized email address: '$email'\n" . $herecurr);
+				# ERROR("BAD_SIGN_OFF",
+				#       "Unrecognized email address: '$email'\n" . $herecurr);
 			} else {
 				my $dequoted = $suggested_email;
 				$dequoted =~ s/^"//;
@@ -3256,11 +3256,11 @@ sub process {
 
 # Check for Gerrit Change-Ids not in any patch context
 		if ($realfile eq '' && !$has_patch_separator && $line =~ /^\s*change-id:/i) {
-			if (ERROR("GERRIT_CHANGE_ID",
-			          "Remove Gerrit Change-Id's before submitting upstream\n" . $herecurr) &&
-			    $fix) {
-				fix_delete_line($fixlinenr, $rawline);
-			}
+			# if (ERROR("GERRIT_CHANGE_ID",
+			#           "Remove Gerrit Change-Id's before submitting upstream\n" . $herecurr) &&
+			#     $fix) {
+			# 	fix_delete_line($fixlinenr, $rawline);
+			# }
 		}
 
 # Check if the commit log is in a possible stack dump
@@ -3411,16 +3411,16 @@ sub process {
 		}
 
 # Check for added, moved or deleted files
-		if (!$reported_maintainer_file && !$in_commit_log &&
-		    ($line =~ /^(?:new|deleted) file mode\s*\d+\s*$/ ||
-		     $line =~ /^rename (?:from|to) [\w\/\.\-]+\s*$/ ||
-		     ($line =~ /\{\s*([\w\/\.\-]*)\s*\=\>\s*([\w\/\.\-]*)\s*\}/ &&
-		      (defined($1) || defined($2))))) {
-			$is_patch = 1;
-			$reported_maintainer_file = 1;
-			WARN("FILE_PATH_CHANGES",
-			     "added, moved or deleted file(s), does MAINTAINERS need updating?\n" . $herecurr);
-		}
+		# if (!$reported_maintainer_file && !$in_commit_log &&
+		#     ($line =~ /^(?:new|deleted) file mode\s*\d+\s*$/ ||
+		#      $line =~ /^rename (?:from|to) [\w\/\.\-]+\s*$/ ||
+		#      ($line =~ /\{\s*([\w\/\.\-]*)\s*\=\>\s*([\w\/\.\-]*)\s*\}/ &&
+		#       (defined($1) || defined($2))))) {
+		# 	$is_patch = 1;
+		# 	$reported_maintainer_file = 1;
+		# 	WARN("FILE_PATH_CHANGES",
+		# 	     "added, moved or deleted file(s), does MAINTAINERS need updating?\n" . $herecurr);
+		# }
 
 # Check for adding new DT bindings not in schema format
 		if (!$in_commit_log &&
@@ -3654,8 +3654,8 @@ sub process {
 			if ($needs_help &&
 			    $help_length < $min_conf_desc_length) {
 				my $stat_real = get_stat_real($linenr, $ln - 1);
-				WARN("CONFIG_DESCRIPTION",
-				     "please write a help paragraph that fully describes the config symbol\n" . "$here\n$stat_real\n");
+				#WARN("CONFIG_DESCRIPTION",
+				#     "please write a help paragraph that fully describes the config symbol\n" . "$here\n$stat_real\n");
 			}
 		}
 
@@ -3730,8 +3730,8 @@ sub process {
 				$compat3 =~ s/\,([a-z]*)[0-9]*\-/\,$1<\.\*>\-/;
 				`grep -Erq "$compat|$compat2|$compat3" $dt_path`;
 				if ( $? >> 8 ) {
-					WARN("UNDOCUMENTED_DT_STRING",
-					     "DT compatible string \"$compat\" appears un-documented -- check $dt_path\n" . $herecurr);
+					# WARN("UNDOCUMENTED_DT_STRING",
+					#      "DT compatible string \"$compat\" appears un-documented -- check $dt_path\n" . $herecurr);
 				}
 
 				next if $compat !~ /^([a-zA-Z0-9\-]+)\,/;
@@ -3774,10 +3774,10 @@ sub process {
 					     "Missing or malformed SPDX-License-Identifier tag in line $checklicenseline\n" . $herecurr);
 				} elsif ($rawline =~ /(SPDX-License-Identifier: .*)/) {
 					my $spdx_license = $1;
-					if (!is_SPDX_License_valid($spdx_license)) {
-						WARN("SPDX_LICENSE_TAG",
-						     "'$spdx_license' is not supported in LICENSES/...\n" . $herecurr);
-					}
+					#if (!is_SPDX_License_valid($spdx_license)) {
+					#	WARN("SPDX_LICENSE_TAG",
+					#	     "'$spdx_license' is not supported in LICENSES/...\n" . $herecurr);
+					#}
 					if ($realfile =~ m@^Documentation/devicetree/bindings/@ &&
 					    $spdx_license !~ /GPL-2\.0(?:-only)? OR BSD-2-Clause/) {
 						my $msg_level = \&WARN;
@@ -3800,8 +3800,8 @@ sub process {
 
 # check for embedded filenames
 		if ($rawline =~ /^\+.*\b\Q$realfile\E\b/) {
-			WARN("EMBEDDED_FILENAME",
-			     "It's generally not useful to have the filename in the file\n" . $herecurr);
+			#WARN("EMBEDDED_FILENAME",
+			#     "It's generally not useful to have the filename in the file\n" . $herecurr);
 		}
 
 # check we are in a valid source file if not then ignore this hunk
@@ -3964,34 +3964,34 @@ sub process {
 		}
 
 # check multi-line statement indentation matches previous line
-		if ($perl_version_ok &&
-		    $prevline =~ /^\+([ \t]*)((?:$c90_Keywords(?:\s+if)\s*)|(?:$Declare\s*)?(?:$Ident|\(\s*\*\s*$Ident\s*\))\s*|(?:\*\s*)*$Lval\s*=\s*$Ident\s*)\(.*(\&\&|\|\||,)\s*$/) {
-			$prevline =~ /^\+(\t*)(.*)$/;
-			my $oldindent = $1;
-			my $rest = $2;
-
-			my $pos = pos_last_openparen($rest);
-			if ($pos >= 0) {
-				$line =~ /^(\+| )([ \t]*)/;
-				my $newindent = $2;
-
-				my $goodtabindent = $oldindent .
-					"\t" x ($pos / $tabsize) .
-					" "  x ($pos % $tabsize);
-				my $goodspaceindent = $oldindent . " "  x $pos;
-
-				if ($newindent ne $goodtabindent &&
-				    $newindent ne $goodspaceindent) {
-
-					if (CHK("PARENTHESIS_ALIGNMENT",
-						"Alignment should match open parenthesis\n" . $hereprev) &&
-					    $fix && $line =~ /^\+/) {
-						$fixed[$fixlinenr] =~
-						    s/^\+[ \t]*/\+$goodtabindent/;
-					}
-				}
-			}
-		}
+		# if ($perl_version_ok &&
+		#     $prevline =~ /^\+([ \t]*)((?:$c90_Keywords(?:\s+if)\s*)|(?:$Declare\s*)?(?:$Ident|\(\s*\*\s*$Ident\s*\))\s*|(?:\*\s*)*$Lval\s*=\s*$Ident\s*)\(.*(\&\&|\|\||,)\s*$/) {
+		# 	$prevline =~ /^\+(\t*)(.*)$/;
+		# 	my $oldindent = $1;
+		# 	my $rest = $2;
+
+		# 	my $pos = pos_last_openparen($rest);
+		# 	if ($pos >= 0) {
+		# 		$line =~ /^(\+| )([ \t]*)/;
+		# 		my $newindent = $2;
+
+		# 		my $goodtabindent = $oldindent .
+		# 			"\t" x ($pos / $tabsize) .
+		# 			" "  x ($pos % $tabsize);
+		# 		my $goodspaceindent = $oldindent . " "  x $pos;
+
+		# 		if ($newindent ne $goodtabindent &&
+		# 		    $newindent ne $goodspaceindent) {
+
+		# 			if (CHK("PARENTHESIS_ALIGNMENT",
+		# 				"Alignment should match open parenthesis\n" . $hereprev) &&
+		# 			    $fix && $line =~ /^\+/) {
+		# 				$fixed[$fixlinenr] =~
+		# 				    s/^\+[ \t]*/\+$goodtabindent/;
+		# 			}
+		# 		}
+		# 	}
+		# }
 
 # check for space after cast like "(int) foo" or "(struct foo) bar"
 # avoid checking a few false positives:
@@ -6920,8 +6920,8 @@ sub process {
 						if (!defined($stat_real)) {
 							$stat_real = get_stat_real($linenr, $lc);
 						}
-						WARN("VSPRINTF_SPECIFIER_PX",
-						     "Using vsprintf specifier '\%px' potentially exposes the kernel memory layout, if you don't really need the address please consider using '\%p'.\n" . "$here\n$stat_real\n");
+						# WARN("VSPRINTF_SPECIFIER_PX",
+						#      "Using vsprintf specifier '\%px' potentially exposes the kernel memory layout, if you don't really need the address please consider using '\%p'.\n" . "$here\n$stat_real\n");
 					}
 				}
 				if ($bad_specifier ne "") {
@@ -7204,14 +7204,14 @@ sub process {
 		}
 
 # checks for new __setup's
-		if ($rawline =~ /\b__setup\("([^"]*)"/) {
-			my $name = $1;
-
-			if (!grep(/$name/, @setup_docs)) {
-				CHK("UNDOCUMENTED_SETUP",
-				    "__setup appears un-documented -- check Documentation/admin-guide/kernel-parameters.txt\n" . $herecurr);
-			}
-		}
+		# if ($rawline =~ /\b__setup\("([^"]*)"/) {
+		# 	my $name = $1;
+
+		# 	if (!grep(/$name/, @setup_docs)) {
+		# 		CHK("UNDOCUMENTED_SETUP",
+		# 		    "__setup appears un-documented -- check Documentation/admin-guide/kernel-parameters.txt\n" . $herecurr);
+		# 	}
+		# }
 
 # check for pointless casting of alloc functions
 		if ($line =~ /\*\s*\)\s*$allocFunctions\b/) {
diff --git a/scripts/kconfig/Makefile b/scripts/kconfig/Makefile
index fb50bd4f4103..fc6e0b0b1536 100644
--- a/scripts/kconfig/Makefile
+++ b/scripts/kconfig/Makefile
@@ -99,7 +99,14 @@ else
 endif
 
 %_defconfig: $(obj)/conf
-	$(Q)$< $(silent) --defconfig=arch/$(SRCARCH)/configs/$@ $(Kconfig)
+# config cannot be used here to mark AMLOGIC modifications
+# support two config directory, arch/$(SRCARCH)/configs/ and $(COMMON_DRIVERS_DIR)/arch/$(SRCARCH)/configs/
+	$(if $(wildcard $(srctree)/$(COMMON_DRIVERS_DIR)/arch/$(SRCARCH)/configs/$@), \
+		$(Q)$< $(silent) --defconfig=$(COMMON_DRIVERS_DIR)/arch/$(SRCARCH)/configs/$@ $(Kconfig), \
+		$(Q)$< $(silent) --defconfig=arch/$(SRCARCH)/configs/$@ $(Kconfig))
+# else
+#	$(Q)$< $(silent) --defconfig=arch/$(SRCARCH)/configs/$@ $(Kconfig)
+# endif
 
 configfiles = $(wildcard $(srctree)/kernel/configs/$(1) $(srctree)/arch/$(SRCARCH)/configs/$(1))
 all-config-fragments = $(call configfiles,*.config)
diff --git a/usr/include/Makefile b/usr/include/Makefile
index 58173cfe5ff1..f442dc4a4a51 100644
--- a/usr/include/Makefile
+++ b/usr/include/Makefile
@@ -8,6 +8,10 @@
 # We cannot go as far as adding -Wpedantic since it emits too many warnings.
 UAPI_CFLAGS := -std=c90 -Wall -Werror=implicit-function-declaration
 
+ifdef CONFIG_AMLOGIC_DRIVER
+UAPI_CFLAGS += -Werror
+endif
+
 # In theory, we do not care -m32 or -m64 for header compile tests.
 # It is here just because CONFIG_CC_CAN_LINK is tested with -m32 or -m64.
 UAPI_CFLAGS += $(filter -m32 -m64 --target=%, $(KBUILD_CPPFLAGS) $(KBUILD_CFLAGS))
-- 
2.50.1

